lib/lsif_graph.dart

```diff
@@ -187,38 +207,80 @@ class Metadata extends Element {
 
   String projectRoot;
 
-  Metadata(this.projectRoot) : super(id: 'meta');
+  Metadata(this.projectRoot);
 
   @override
   Map<String, Object> toLsif() => {
         ...super.toLsif(),
         'projectRoot': projectRoot,
-        'version': '0.5.0',
+        'version': '0.4.3',
         'positionEncoding': 'utf-16',
         'toolInfo': toolInfo,
       };
 
   Map<String, Object> get toolInfo => {
-        'name': 'simple_lsif',
+        'name': 'lsif_indexer',
         'args': [],
         'version': 'dev',
       };
 }
 
-class Contains extends Edge {
+/// A "contains" edge specifically for a [Document], which contains references and declarations.
+class DocumentContains extends Edge {
   @override
   String get label => 'contains';
 
-  Contains(this.container);
+  DocumentContains(this.container);
 
   Document container;
 
   @override
   Map<String, Object> toLsif() =>
       {...super.toLsif(), 'outV': container.jsonId, 'inVs': incomingEdges};
 
-  List<String> get incomingEdges => [
-        ...container.references,
-        ...container.declarations,
-      ].map((each) => each.range.jsonId).toList();
+  // TODO: Tidy this up, and the semantics of when things are added to declarations. See
+  // emitReferencesAndDeclarations in Document, where we push things into the references
+  // from declarations, but not vice versa.
+  List<String> get incomingEdges => {
+        for (var ref in container.references) ref.range.jsonId,
+        for (var ref in container.externalReferences) ref.range.jsonId,
+        for (var declaration in container.declarations)
+          declaration.range.jsonId,
+        for (var ref in container.references) ref.declaration.range.jsonId,
+      }.toList();
+}
+
+/// A "contains" edge specifically for a [Project], which contains [Document]s
+class ProjectContains extends Edge {
+  @override
+  String get label => 'contains';
+
+  ProjectContains(this.project);
+
+  Project project;
```

Comment:
#nit I'd consider making instance variables `private`/`final` where possible, I find it helps determining things like:
- Where they're set from (is it just the constructor, or can other objects mutate it?)
- Is it used at all? (The analyzer picks up on this if it is private, but not public)
- Is it ever modified after construction?

Not that it is necessarily important for this code, since I don't think it intends to be a library, but I think `not final` -> `final` is a major change, and `public` -> `private` is major also (assuming they're exported) where the other way around is not. So if we're not externally mutating it currently and decide we need to, it would be easy to make it public or remove `final`

Reply:
That's not really a nit. Updated issues #11 and #12 to call out those points. I've definitely erred so far on the side of making things public, since I've just been hacking it out, and this isn't really intended as a library, at least at this point. But we probably do want to sort that out some, and it should be done in the context of splitting up the current huge libraries.

I'm not a huge fan of final, but these are essentially immutable entities once constructed, and it would probably make sense in a lot of these places. It also would fit in with storing entity relationships rather than their IDs. Might be a bit tricky with graph construction, but worth looking at.